tests/python_client/testcases/test_query.py

```diff
@@ -686,6 +686,141 @@ def test_query_expr_json_contains_pagination(self, enable_dynamic_field):
+                           check_task=CheckTasks.check_query_results, check_items={exp_res: res})
+
+    @pytest.mark.tags(CaseLabel.L2)
+    @pytest.mark.parametrize("offset", [100, 1000])
+    @pytest.mark.parametrize("limit", [100, 1000])
+    def test_query_expr_empty_with_random_pk(self, limit, offset):
+        """
+        target: test query with empty expression
+        method: create a collection using random pk, query empty expression with a limit
+        expected: return topK results by order
+        """
+        # 1. initialize with data
+        collection_w = self.init_collection_general(prefix, with_json=False)[0]
+
+        # 2. generate unordered pk array and insert
+        unordered_ids = [i for i in range(ct.default_nb)]
+        random.shuffle(unordered_ids)
+        float_value = [np.float32(i) for i in unordered_ids]
+        string_value = [str(i) for i in unordered_ids]
+        vector_value = cf.gen_vectors(nb=ct.default_nb, dim=ct.default_dim)
+        collection_w.insert([unordered_ids, float_value, string_value, vector_value])
+        collection_w.load()
+
+        # 3. query with empty expr and check the result
+        exp_ids, res = sorted(unordered_ids)[:limit], []
+        for ids in exp_ids:
+            res.append({ct.default_int64_field_name: ids, ct.default_string_field_name: str(ids)})
+
+        collection_w.query("", limit=limit, output_fields=[ct.default_string_field_name],
+                           check_task=CheckTasks.check_query_results, check_items={exp_res: res})
+
+        # 4. query with pagination
+        exp_ids, res = sorted(unordered_ids)[:limit + offset][offset:], []
+        for ids in exp_ids:
+            res.append({ct.default_int64_field_name: ids, ct.default_string_field_name: str(ids)})
+
+        collection_w.query("", limit=limit, offset=offset, output_fields=[ct.default_string_field_name],
+                           check_task=CheckTasks.check_query_results, check_items={exp_res: res})
+
+    @pytest.mark.tags(CaseLabel.L0)
+    def test_query_expr_with_limit_offset_out_of_range(self):
+        """
+        target: test query with empty expression
+        method: query empty expression with limit and offset out of range
+        expected: raise error
+        """
+        # 1. initialize with data
+        collection_w = self.init_collection_general(prefix, True)[0]
+
+        # 2. query with limit > 16384
+        error = {ct.err_code: 1, ct.err_msg: "invalid max query result window, (offset+limit) should be in range [1, 16384]"}
+        collection_w.query("", limit=16385, check_task=CheckTasks.err_res, check_items=error)
+
+        # 3. query with offset + limit > 16384
+        collection_w.query("", limit=1, offset=16384, check_task=CheckTasks.err_res, check_items=error)
+
+        # 4. query with limit < 0
+        error = {ct.err_code: 1, ct.err_msg: "invalid max query result window, offset [-1] is invalid, should be gte than 0"}
```

Comment:
This error message does not show what is the error, should be greater than 0 , while 16384 is indeed greater than 0, should be raise an issue here.

Reply:
Here limit=1, offset=16384 will raise error: invalid max query result window, (offset+limit) should be in range [1, 16384]

limit=2, offset=-1 will raise error invalid max query result window, offset [-1] is invalid, should be gte than 0

So it is valid